### ISSUE TITLE
False positive: String vs. slice comparison

### ISSUE THREAD

**Issue by lancelote** [intellij-rust/intellij-rust#4224]
## Environment

* **IntelliJ Rust plugin version:** 0.2.103.2699-192-nightly
* **Rust toolchain version:** rustc 1.36.0 (a53f9df32 2019-07-03)
* **IDE name and version:** CLion 2019.3 Nightly 193.1888
* **Operating system:** macOS 10.14.4

## Problem description

I have the following code:

```rust
fn main() {
    if String::from("hello") == "hello" {
        println!("works 1");
    }
    if "hello" == String::from("hello") {
        println!("works 2");
    }
}
```

Plugin highlights the slice in the first comparison as an error and reports:

> expected `String`, found `&str`

The code is working though. Do I missing something or is it a false positive?

**Comment by Aloso** [comment]
This only affects the _default macro engine_.

There are `PartialEq` implementations for comparing a `String` with a string slice. It's defined [here](https://doc.rust-lang.org/src/alloc/string.rs.html#1838-1863):

```rust
macro_rules! impl_eq {
    ($lhs:ty, $rhs: ty) => {
        #[stable(feature = "rust1", since = "1.0.0")]
        impl<'a, 'b> PartialEq<$rhs> for $lhs {
            #[inline]
            fn eq(&self, other: &$rhs) -> bool { PartialEq::eq(&self[..], &other[..]) }
            #[inline]
            fn ne(&self, other: &$rhs) -> bool { PartialEq::ne(&self[..], &other[..]) }
        }

        #[stable(feature = "rust1", since = "1.0.0")]
        impl<'a, 'b> PartialEq<$lhs> for $rhs {
            #[inline]
            fn eq(&self, other: &$lhs) -> bool { PartialEq::eq(&self[..], &other[..]) }
            #[inline]
            fn ne(&self, other: &$lhs) -> bool { PartialEq::ne(&self[..], &other[..]) }
        }

    }
}

impl_eq! { String, str }
impl_eq! { String, &'a str }
impl_eq! { Cow<'a, str>, str }
impl_eq! { Cow<'a, str>, &'b str }
impl_eq! { Cow<'a, str>, String }
```

But when expanding with the default macro engine, IntelliJ Rust doesn't find these implementations.

**Comment by Undin** [comment]
The reason of problem that the corresponding `impl` for `String` and `&str` comparison is generated by macro in stdlib. And default (at this moment) macro expansion engine can't process generated `impl` blocks and `mod` declarations for some technical reasons.
To fix it try to enable new experimental macro expansion engine - it should fix this issue. See #3628 for more details
